### ISSUE TITLE
[Messenger] [DoctrineMessenger] "Deadlock found when trying to get lock" when DELETEing from messages table by delivered_at

### ISSUE THREAD

**Issue by manuelkiessling** [symfony/symfony#47366]
### Symfony version(s) affected

5.4.11

### Description

I believe this is not a completely new issue. See #41541 for an older issue that reported the same problem.

We are using Symfony Messenger on top of a MariaDB 10.5.13 table, using only one queue that regularly holds several dozen to a couple hundred messages with the same available_at date, and around 20 consumers/workers that listen on this queue in parallel.

We thus have a high level of parallelization when the available_at time comes (which is once per hour).

This always results in a couple of `"An exception occurred while executing 'DELETE FROM symfony_messages WHERE delivered_at = ?' with params ["9999-12-31"]: SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction"` exceptions.

The solution for the previously reported problem was to not set indices on table fields `queue_name` and `available_at` - see #42345.

However, this was later taken back by #45888, which instead uses a "soft-delete" approach to avoid deadlocks.

Alas, I have to report that deadlocks still occur when multiple consumers consume the same queue in parallel.

### How to reproduce

The reproducer from #41541 should do the job.

**Comment by punk-raider** [comment]
W have a similar problem when adding messages: 
```
SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction
An exception occurred while executing 'INSERT INTO messenger_messages (...
```

**Comment by carsonbot** [comment]
Hey, thanks for your report!
There has not been a lot of activity here for a while. Is this bug still relevant? Have you managed to find a workaround?

**Comment by hebrianvasyl** [comment]
Hello,

I noticed the same issue of "Deadlock found" on Symfony 6.2. It occurs during DELETE, UPDATE and SELECT...FOR UPDATE queries.

What's interesting is that it was working fine initially, but at some point, it started happening. Perhaps it's due to increased site usage.

I have two queues and four supervisor jobs running, which handle both queues:
`bin/console messenger:consume async async_priority_low --time-limit=3600`

**Comment by dpfaffenbauer** [comment]
We have the same problem, we run multiple workers for multiple queues, and with high load we get lock errors. Only solution is to restart the workers and release the locks.

**Comment by carsonbot** [comment]
Hey, thanks for your report!
There has not been a lot of activity here for a while. Is this bug still relevant? Have you managed to find a workaround?

**Comment by carsonbot** [comment]
Friendly ping? Should this still be open? I will close if I don't hear anything.

**Comment by dpfaffenbauer** [comment]
still happens

**Comment by dpi** [comment]
Might be resolved with #54606

**Comment by dpfaffenbauer** [comment]
we solved it with multiple tables for different transports.

**Comment by bobvandevijver** [comment]
We started seeing this after upgrading from MariaDB 10.3 (Ubuntu 20.04) to 10.11 (Ubuntu 24.04). We also see other performance issues due to this update, indicating that something changed in the optimisation of MariaDB regarding queries.

Edit; in our case it seems that the default `innodb_buffer_pool_size` was simply not sufficient due to large messages in the queue, but we have also switched to multiple tables for different transports.

**Comment by netbrothers-tr** [comment]
We are seeing this

- with low volume,
- only a single transport and
- despite #54606.

We are unable to figure out, why it happens. So far, only one of multiple applications on the same machine is affected.

**Comment by bobvandevijver** [comment]
@netbrothers-tr Are the records large? We solved it by enlarging the innodb buffer pool size.

**Comment by cancan101** [comment]
I am seeing deadlock as well, might be related to the select for update and perhaps when errors are thrown in message handling https://github.com/symfony/symfony/issues/20619#issuecomment-2918089196

**Comment by psihius** [comment]
https://github.com/symfony/symfony/pull/61963 is properly fixing this issue.

**Comment by netbrothers-tr** [comment]
> [@netbrothers-tr](https://github.com/netbrothers-tr) Are the records large? We solved it by enlarging the innodb buffer pool size.

For the sake of completeness: the entries came from the mailer/messenger components so they were e-mails (in our case without attachments, but with a small PNG logo image), so the answer is probably "no" ?! Since AFAIK the problem didn't come up again, it seems to be gone for good. :)

**Comment by Toflar** [comment]
Which by the way is precisely the reason for https://github.com/symfony/symfony/issues/60390 to exist 😇

**Comment by netbrothers-tr** [comment]
@Toflar, I love your enthusiasm for #60390 and I strongly support it, but in our case that wasn't the issue, since we were well under (!) the limits of at least RabbitMQ and Kafka. 😉 However, you are totally right about the design flaw.
